### ISSUE TITLE
Incorrect date change at local midnight in TimestampFormatter for non-GMT timezones

### ISSUE THREAD

**Issue by tarun-t** [odygrd/quill#498]
There is a bug in the ```StringFromTime``` class that affects timestamp formatting for local timezones, particularly when crossing midnight. The current implementation does not correctly handle date changes at local midnight for time zones other than GMT.

Note on reproduction:
Reproducing this issue is tricky because the ```StringFromTime::init``` method uses the current system time when it's called. This means that the behavior of the logger will vary depending on the time of day when the code is executed. To consistently reproduce the issue, you would need to mock the system time or add a way to inject a specific timestamp for initialization and testing purposes.

The issue stems from the different handling of GMT and local timezones in the StringFromTime class. For local timezones, it recalculates every hour without special handling for midnight.

**Comment by odygrd** [comment]
Hey thanks a lot for reporting this. By looking at the code i couldn’t figure out what is wrong. I will try to reproduce it over the new few days.

for local time the init method takes the current timestamp, rounds it down to the sharp hour and then adds one more hour for the recalculation. 
For example if you start it at 13:30, then it should recalculate at 14:00,15:00,etc,..
That means the midnight is also included as we are recalculating every hour

**Comment by odygrd** [comment]
Using this example and mocking the system time forward and backward i could not preproduce

<img width="1013" alt="image" src="https://github.com/user-attachments/assets/b8252f9c-9a58-4290-aa90-773e08d9f019">

**Comment by tarun-t** [comment]
I could construct this small example that illustrates the behavior I was talking about. Here's the piece of code:

```
#include "quill/backend/StringFromTime.h"
#include <chrono>
#include <climits>
#include <iostream>
#include <string>

int main()
{
  quill::detail::StringFromTime formatter;
  formatter.init("%Y-%m-%d %H:%M:%S", quill::Timezone::LocalTime);

  using namespace std::chrono;

  // Get current local time
  auto now = zoned_time{ current_zone(), system_clock::now() };

  // Get local midnight tomorrow
  auto midnight_tomorrow = floor<days>(now.get_local_time()) + days{ 1 };

  // Convert to a time_point
  auto local_midnight = local_days{ midnight_tomorrow };

  // Convert to system_clock::time_point
  auto sys_time = zoned_time{ current_zone(), local_midnight }.get_sys_time();

  // Get the epoch timestamp
  auto local_secs = duration_cast<seconds>(sys_time.time_since_epoch());

  std::cout << "Epoch timestamp for midnight tomorrow: " << local_secs.count() << " seconds\n";

  auto first = formatter.format_timestamp(local_secs.count() - 1); 

  auto second = formatter.format_timestamp(local_secs.count() + 100);

  std::cout << "First: " << first << " Second: " << second << '\n';
  return 0;
}
```

When I run it today on a machine with the time zone set to IST, here's the output of the program:
```
Epoch timestamp for midnight tomorrow: 1721413800 seconds
First: 2024-07-19 23:59:59 Second: 2024-07-19 24:01:40
```

**Comment by odygrd** [comment]
thanks so much for reporting this and for the detailed explanation, I could reproduce it. It should be fixed in 4df7dc1b9e077312460aacf82f19944cdecb83e1 

If you notice any other issue just let me know : )
